cal-itp/data-infra
# feat(amplitude-ingest): Amplitude to Airflow task

# Overall Description
Related to #960, closes #1046 

Merges this PR into `feat-amplitude-operator`

This PR adds to the `amplitude_to_df` function some logic to call the Amplitude Export API and read the returned zipped stream into a Pandas DataFrame. 

## What this PR does and does not do
- See the first commit [`98403dd`](https://github.com/cal-itp/data-infra/pull/1081/commits/98403ddfb5f402225407f9753651b5c67f8c933a) for the basic implementation
- This PR does not add error handling.
- This PR does not attempt to transform the raw data into any specific schema.
- This PR does not have Pytests. (We could write proper pytests which I think would be done as a separate issue.)


## Pre-requisites for testing locally

1. Change `.devcontainer/Dockerfile` line 4:
```diff
-ARG VARIANT="3.9"
+ARG VARIANT="3.8"
```
2. Export `CALITP_AMPLITUDE_BENEFITS_API_KEY` and `CALITP_AMPLITUDE_BENEFITS_SECRET_KEY` key value pairs. Get key/secret from Amplitude.


## How to test

I tested these changes myself by writing some simple, throwaway code to call the function with various start/end dates and keys I got from Amplitude. 

1. Modify `airflow/debug.py`
```diff
-import os
-import sys
+# import os
+# import sys
 
-import dotenv
+# import dotenv
 
-dotenv.load_dotenv()
+# dotenv.load_dotenv()
 
-sys.path.append(os.path.abspath("plugins"))
+# sys.path.append(os.path.abspath("plugins"))
+
+import plugins.operators.amplitude_to_warehouse as atw
+
+df = atw.amplitude_to_df("20220101T05", "20220102T05")
+print(df)
```
2. Run `python -m debug.py` and expect data to be returned:
```bash
       app                             device_id  ... $schema             processed_time
0   304110  2842c0b8-a104-847e-63b7-dab84adb1196  ...      13 2022-01-01 05:07:43.196789
1   304110  9bc14448-275d-9755-9425-61aed7b38605  ...      13 2022-01-01 05:03:55.122934
2   304110  cb836441-943a-a9a4-baaf-8f28e9603d62  ...      13 2022-01-01 05:08:51.600550
3   304110  92d55c52-da91-ab9f-957e-1bf327d915e1  ...      13 2022-01-01 05:10:04.664107
```


## Checklist for all PRs

- [x] Run `pre-commit run --all-files` to make sure markdown/lint passes
- [x] Link this pull request to all issues that it will close using keywords (see GitHub docs about [Linking a pull request to an issue using a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). Also mention any issues that are partially addressed or are related.
